Fix and test for issue #4117

rwgk opened the pull request:
<!--
Title (above): please place [branch_name] at the beginning if you are targeting a branch other than master. *Do not target stable*.
It is recommended to use conventional commit format, see conventionalcommits.org, but not required.
-->
## Description
See #3861 and #4117.

<!-- Include relevant issues or PRs here, describe what changed and why -->


## Suggested changelog entry:

<!-- Fill in the below block with the expected RestructuredText entry. Delete if no entry needed;
     but do not delete header or rst block if an entry is needed! Will be collected via a script. -->

```rst
TODO
```

<!-- If the upgrade guide needs updating, note that here too -->

rwgk commented:
@laramiel thanks for suggesting the test!

(I wasn't getting very lucky looking under https://github.com/wassimj/Topologic)

rwgk commented:
@henryiii there is a test here.

rwgk commented:
@Lalaland @laramiel It's interesting that the CI succeeded for #4273, it simply replaces `intrinsic_t<T>` with `T`.

@Lalaland did you come to the idea that `intrinsic_t` is unnecessary by inspecting the calling code?

I almost think it will be best to drop this PR and reopen and merge #4273 (the only hiccup was the unused ctypes import). Although it might be useful to add in the Python part of the test from here, which has a Python override returning `self` (@laramiel's idea). I believe it is redundant for the purpose of exercising the exact core fix, but it is a different situation that we evidently don't cover otherwise, and is very little extra code.

@henryiii for awareness.

BTW: Amazing coincidence that 3 people worked on fixes in 2 PRs at the same time.

Lalaland commented:
> @Lalaland <https://github.com/Lalaland> did you come to the idea that
intrinsic_t is unnecessary by inspecting the calling code?

I switched it to T just to revert to the old behavior in 2.9

Your version with remove_cv is more correct (and an improvement over 2.9,
as opposed to just fixing the regression).

I can quickly reopen #4273 with the needed changes (remove_cv and returning
self).

On Sat, Oct 22, 2022, 12:23 AM Ralf W. Grosse-Kunstleve <
***@***.***> wrote:

> @Lalaland <https://github.com/Lalaland> @laramiel
> <https://github.com/laramiel> It's interesting that the CI succeeded for
> #4273 <https://github.com/pybind/pybind11/pull/4273>, it simply replaces
> intrinsic_t<T> with T.
>
> @Lalaland <https://github.com/Lalaland> did you come to the idea that
> intrinsic_t is unnecessary by inspecting the calling code?
>
> I almost think it will be best to drop this PR and reopen and merge #4273
> <https://github.com/pybind/pybind11/pull/4273> (the only hiccup was the
> unused ctypes import). Although it might be useful to add in the Python
> part of the test from here, which has a Python override returning self (
> @laramiel <https://github.com/laramiel>'s idea). I believe it is
> redundant for the purpose of exercising the exact core fix, but it is a
> different situation that we evidently don't cover otherwise, and is very
> little extra code.
>
> @henryiii <https://github.com/henryiii> for awareness.
>
> BTW: Amazing coincidence that 3 people worked on fixes in 2 PRs at the
> same time.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pybind/pybind11/pull/4274#issuecomment-1287658786>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACTRWJ5Z633JHZPSANCH4DWEOJAPANCNFSM6AAAAAARLU5JRY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Lalaland commented:
Don't actually have permission to reopen PRs, so created #4275

Skylion007 commented on tests/test_virtual_functions.cpp:
```diff
@@ -230,6 +230,35 @@ inline int test_override_cache(std::shared_ptr<test_override_cache_helper> const
+
+class Animal {
+public:
+    virtual ~Animal() {}
```
```suggestion
    virtual ~Animal() = default;
```

rwgk commented:
Closing, superseded by #4275

rwgk commented:
FWIW / future reference: I rebased and squashed my branch after #4275 was merged, JIC we want to resurrect the test later. I also verified that the test succeeds as expected.
